### DISCUSSION TITLE
Explicit EOI token needed for empty input in some cases?

### DISCUSSION THREAD

**Discussion by carueda** [zesterer/chumsky#573] [Q&A]
Trying to better understand the error reporting for empty input in general.

(Using chumsky1.0.0-alpha.6)

Based on this simple token type and parser:

```rust
#[derive(Clone, Debug, PartialEq)]
enum Token<'src> {
    Ident(&'src str),
    Gets(&'src str),
    Eoi, // for the question here
}

fn program_parser<'tokens, 'src: 'tokens>(
) -> impl Parser<
  'tokens,
   &'tokens [Token<'src>],
   (),
   extra::Err<Rich<'tokens, Token<'src>>>
>{
    let ident = select! { Token::Ident(_) }.labelled("ident");
    let gets = select! { Token::Gets(_) }.labelled("=");
    let expr = select! { Token::Ident(_) }.labelled("expression");

    group((ident, gets, expr)).to(()).labelled("program")
}
```

With `Token::Eoi` as an explicit token in the input sequence, that is, `vec![Token::Eoi]`, I seem to get the desirable behavior:

```rust
    let tokens = &vec![Token::Eoi];
    let (program, errs) = program_parser().parse(tokens).into_output_errors();

    assert_eq!(program, None);
    assert_eq!(errs.len(), 1);
    let rich = &errs[0];
    match rich.reason() {
      RichReason::ExpectedFound { expected, found } => {
        assert_eq!(found, &Some(Maybe::Val(Token::Eoi)));
        assert_eq!(expected, &[RichPattern::Label("program")]);
      },
      r => panic!("unexpected reason: {:#?}", r),
    }
```
that is (basically), with found=Eoi and expected="program".

However, why not a similar behavior for `expected` when `vec![]` is given as input:

```rust
    let tokens = &vec![];
    let (program, errs) = program_parser().parse(tokens).into_output_errors();
    assert_eq!(program, None);
    assert_eq!(errs.len(), 1);
    let rich = &errs[0];
    match rich.reason() {
      RichReason::ExpectedFound { expected, found } => {
        assert_eq!(found, &None);
        assert_eq!(expected, &[RichPattern::Label("program")]);
      },
      r => panic!("unexpected reason: {:#?}", r),
    }
```
- the `assert_eq!(found, &None)` in this case (which passes) seems appropriate as `found` is in terms of tokens, and there are no tokens at all in the input;
- however, I think the assert on `expected` should still pass as indicated, but fails because `expected` turns out to be empty:
```
Diff < left / right > :
<[]
>[
>    "program",
>]
```

I'd appreciate any comments/clarifications.

**Comment by zesterer** [comment]
Yes, I think you're right that the labelling behaviour in this case is inconsistent. I think the problem here is that the parser is assuming that because it couldn't even *start* parsing a `program`, it should fall back to another rule: but there aren't any. I think this may be relevant to #518 and probably needs a revisit.

That aside, I would advise not treating labelling as a precise, documented behaviour: exactly how chumsky does error prioritisation and generation is deliberately considered to be an implementation issue so that it can be improved incrementally without the change being considered breaking.

**Comment by carueda** [comment]
Actually –and sorry for any confusion– I used labelled definitions just for illustration, but the core question is, why `expected` turns out empty when the given input is empty?  Thanks again.
